### PR TITLE
Fixes #44

### LINKED ISSUE #44: `plot.grid_search` throws a `TypeError` when parameters cannot be sorted.
# Describe the bug

`plot.grid_search` throws a `TypeError` when parameters cannot be sorted.

# Steps/Code to Reproduce 

 ```python
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn import datasets

from sklearn_evaluation.plot import grid_search

iris = datasets.load_iris()

std = StandardScaler()
est = KNeighborsClassifier(n_neighbors=5)

pipe = Pipeline([
    ('std', std),
    ('est', est)
])

parameters = {
   'std': [None, std],
   'est__n_neighbors': [1, 10, 50, 100],
}

clf = GridSearchCV(pipe, parameters, cv=5)

X, y = datasets.make_classification(1000, 10, n_informative=5, class_sep=0.7)
clf.fit(X, y)

# changing numeric parameter without any restrictions
# in the rest of the parameter set
grid_search(clf.cv_results_, change='est__n_neighbors')
plt.show()
```

# Expected Results

Should display Grid search results plot with legend :

std: None
std: StandardScaler()

# Actual Results

```python-traceback
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/home/adrien/Dropbox/workspace/sklearn-evaluation/issues/issue_43.py in <cell line: 23>()
     [30](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/issues/issue_43.py?line=29) clf.fit(X, y)
     [32](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/issues/issue_43.py?line=31) # changing numeric parameter without any restrictions
     [33](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/issues/issue_43.py?line=32) # in the rest of the parameter set
---> [34](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/issues/issue_43.py?line=33) grid_search(clf.cv_results_, change='est__n_neighbors')
     [35](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/issues/issue_43.py?line=34) plt.show()

File ~/Dropbox/workspace/sklearn-evaluation/issues/../src/sklearn_evaluation/plot/grid_search.py:81, in grid_search(cv_results_, change, subset, kind, cmap, ax)
     [74](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=73) grid_scores = [
     [75](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=74)     gs(p, m, s) for p, m, s in
     [76](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=75)     zip(cv_results_['params'], cv_results_['mean_test_score'],
     [77](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=76)         cv_results_['std_test_score'])
     [78](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=77) ]
     [80](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=79) if isinstance(change, string_types) or len(change) == 1:
---> [81](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=80)     return _grid_search_single(grid_scores, change, subset, kind, ax)
     [82](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=81) elif len(change) == 2:
     [83](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=82)     return _grid_search_double(grid_scores, change, subset, cmap, ax)

File ~/Dropbox/workspace/sklearn-evaluation/issues/../src/sklearn_evaluation/plot/grid_search.py:132, in _grid_search_single(grid_scores, change, subset, kind, ax)
    [126](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=125)     # bar shifter is just a wrapper around matplotlib's bar plot
    [127](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=126)     # to automatically calculate the left position on each bar
    [128](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=127)     bar_shifter = BarShifter(g_number=change_unique,
    [129](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=128)                              g_size=len(groups),
    [130](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=129)                              ax=ax)
--> [132](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=131) for params_kv, group in _sorted_map_iter(groups):
    [133](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=132)     # get the x and y values for each grid_score on this group
    [134](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=133)     # also calculate the std
    [135](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=134)     x = [element.parameters[change] for element in group]
    [136](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/plot/grid_search.py?line=135)     y = [element.mean_validation_score for element in group]

File ~/Dropbox/workspace/sklearn-evaluation/issues/../src/sklearn_evaluation/util.py:85, in _sorted_map_iter(d)
     [84](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/util.py?line=83) def _sorted_map_iter(d):
---> [85](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/util.py?line=84)     ord_keys = sorted(d.keys())
     [86](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/util.py?line=85)     for k in ord_keys:
     [87](file:///home/adrien/Dropbox/workspace/sklearn-evaluation/src/sklearn_evaluation/util.py?line=86)         yield (k, d[k])

TypeError: '<' not supported between instances of 'StandardScaler' and 'NoneType'
```

# Versions

```python
import sklearn; sklearn.show_versions()
```

```
System:
    python: 3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:24:11)  [GCC 9.4.0]
executable: /home/adrien/mambaforge/bin/python
   machine: Linux-4.15.0-175-generic-x86_64-with-glibc2.27

Python dependencies:
          pip: 22.0.4
   setuptools: 60.10.0
      sklearn: 1.0.2
        numpy: 1.22.3
        scipy: 1.8.0
       Cython: None
       pandas: 1.4.2
   matplotlib: 3.5.1
       joblib: 1.1.0
threadpoolctl: 3.1.0

Built with OpenMP: True
/home/adrien/mambaforge/lib/python3.9/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
```

### PR DESCRIPTION
**Pull request by adrtod**
And remove prepended commas in legend labels

### CODE REVIEW

**Comment by edublancas** [comment]
thanks for your contribution! I think it'd be better to add this to `_sorted_map_iter`: try to return the sorted keys, and if the TypeError is raised, then skip the sorting, this way we won't break the API. Can you add that change and add some tests?

https://github.com/edublancas/sklearn-evaluation/blob/8676283032880c283f478d282fd3438d7cb30e6f/src/sklearn_evaluation/util.py#L84

**Comment by adrtod** [comment]
I am not sure that you want to sort automatically. This way, you can control the ordering from the ParameterGrid. Sorting can be done by the user if that's needed.

```
est = KNeighborsClassifier()

parameters = {
   'weights': ['uniform', 'distance'],
   'n_neighbors': [1, 10, 50, 100],
}

clf = GridSearchCV(est, parameters, cv=5)

clf.fit(X, y)

grid_search(clf.cv_results_, change='n_neighbors')
plt.show()
```

will display the legend in the order chosen instead of the alphabetic order
weights : 'uniform'
weights : 'distance'

Just change the order by e.g. sorting if needed 
```
parameters = {
   'weights': sorted(['uniform', 'distance']),
   'n_neighbors': [1, 10, 50, 100],
}
```

File: src/sklearn_evaluation/plot/grid_search.py [hunk 1]
```diff
@@ -139,7 +139,7 @@ def _grid_search_single(grid_scores, change, subset, kind, ax):
         # take (param, value) and convert them to 'param: value'
         label = ['{}: {}'.format(*t) for t in params_kv]
         # now convert it to one string
-        label = reduce(lambda x, y: x + ', ' + y, label, '')
+        label = ', '.join(label)
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/grid_search.py:142]: this is breaking the tests, it'll produce a string that starts with `,` - please revert the change
single_numeric_partially_restricted

**Review by edublancas — CHANGES_REQUESTED** [review]
Ok, I agree that letting the user sort is good. To keep backward compatibility, I'd suggest adding a `sort=True` argument to the function, and then if the argument is not sortable, show an error message suggesting to pass `sort=False`

File: src/sklearn_evaluation/plot/grid_search.py (hunk 1 above)
Comment (adrtod) [reply src/sklearn_evaluation/plot/grid_search.py:142]: Ok I will add the `sort` argument.
However my change line 142 for the labeling actually fixes the leading comma problem.

Comment (adrtod) [reply src/sklearn_evaluation/plot/grid_search.py:142]:
```py
from functools import reduce

label = ['criterion: entropy', 'max_features: sqrt']
label_bad = reduce(lambda x, y: x + ', ' + y, label, '')
label_good = ', '.join(label)

print('label_bad =', label_bad)
print('label_good =', label_good)
```
will output
```
label_bad = , criterion: entropy, max_features: sqrt
label_good = criterion: entropy, max_features: sqrt
```

File: issues/issue_44.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+# %%
```
Comment (edublancas) [inline issues/issue_44.py:1]: this looks like a testing file, can you delete it?

File: src/sklearn_evaluation/plot/grid_search.py [hunk 3]
```diff
@@ -18,7 +17,9 @@ def grid_search(cv_results_,
                 subset=None,
                 kind='line',
                 cmap=None,
-                ax=None):
+                ax=None,
+                sort=False,
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/grid_search.py:21]: please switch the default to sort=True to keep backward compatibility

File: src/sklearn_evaluation/util.py [hunk 4]
```diff
@@ -81,8 +81,14 @@ def fn(obj):
...
 
 
-def _sorted_map_iter(d):
-    ord_keys = sorted(d.keys())
+def _sorted_map_iter(d, sort=True):
+    if sort:
+        try:
+            ord_keys = sorted(d.keys())
+        except TypeError:
+            raise TypeError('Sorting not supported. Plase pass sort=False')
```
Comment (edublancas) [inline src/sklearn_evaluation/util.py:89]: python doesn't like when we raise exceptions inside an `except` block, please change this to a chained exception

**Review by edublancas — CHANGES_REQUESTED** [review]
added a few comments, please also add a test here with `sort=False` https://github.com/edublancas/sklearn-evaluation/blob/master/tests/test_grid_search.py

File: src/sklearn_evaluation/plot/grid_search.py (hunk 1 above)
Comment (adrtod) [reply src/sklearn_evaluation/plot/grid_search.py:145]: I updated the baseline images for the tests

**Comment by adrtod** [comment]
I made the changes requested, added tests and updated the images.

**Comment by adrtod** [comment]
thanks for the release!
